src/OmniSharp.Roslyn.CSharp/Services/Completion/CompletionListCache.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Based on dotnet/roslyn:src/Features/LanguageServer/Protocol/Handler/Completion/CompletionListCache.cs
+// CommitID: 62528a6843dc5b21b1a7d399bae814868456e823
+// Original license is MIT
+
+#nullable enable
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Completion
+{
+    /// <summary>
+    /// Caches completion lists in between calls to CompletionHandler and
+    /// CompletionResolveHandler. Used to avoid unnecessary recomputation.
+    /// </summary>
+    internal class CompletionListCache
+    {
+        /// <summary>
+        /// Maximum number of completion lists allowed in cache. Must be >= 1.
+        /// </summary>
+        private const int MaxCacheSize = 2;
```

Comment:
we used to have a single statically cached item, I am not sure I can follow why is there now two? 
is this "just in case"?

Reply:
We now have an after insert step, and the user could technically have triggered a completion session in between. I want to make sure that we handle it somewhat gracefully.